src/v/cluster/tm_stm.cc

```diff
@@ -0,0 +1,401 @@
+    }
+
+    vlog(
+      clusterlog.trace,
+      "Registering existing tx: id={}, pid={}, etag={}",
+      tx_id,
+      pid,
+      etag);
+
+    auto term = _insync_term;
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        co_return tm_stm::op_status::not_found;
+    }
+    if (ptx->second.etag != etag) {
+        co_return tm_stm::op_status::conflict;
+    }
+    auto tx = ptx->second;
+    tx.status = tm_transaction::tx_status::ongoing;
+    tx.pid = pid;
+    tx.tx_seq += 1;
+    tx.partitions = std::vector<tm_transaction::tx_partition>();
+
+    auto r = co_await _tx_locks.find(tx_id)->second->with(
+      [this, term, tx]() { return save_tx(term, tx); });
+
+    if (!r.has_value()) {
+        co_return tm_stm::op_status::unknown;
+    }
+    co_return tm_stm::op_status::success;
+}
+
+ss::future<tm_stm::op_status> tm_stm::register_new_producer(
+  kafka::transactional_id tx_id, model::producer_identity pid) {
+    auto timeout = config::shard_local_cfg().tm_sync_timeout_ms.value();
+    auto is_ready = co_await sync(timeout);
+    if (!is_ready) {
+        co_return tm_stm::op_status::unknown;
+    }
+
+    vlog(clusterlog.trace, "Registering new tx: id={}, pid={}", tx_id, pid);
+
+    auto term = _insync_term;
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx != _tx_table.end()) {
+        co_return tm_stm::op_status::conflict;
+    }
+    if (!_end_locks.contains(tx_id)) {
+        _end_locks.emplace(tx_id, ss::make_lw_shared<mutex>());
+    }
+    auto lock_it = _tx_locks.find(tx_id);
+    if (lock_it == _tx_locks.end()) {
+        _tx_locks.emplace(tx_id, ss::make_lw_shared<mutex>());
+        lock_it = _tx_locks.find(tx_id);
+    }
+
+    co_return co_await lock_it->second->with([this, term, tx_id, pid]() {
+        return register_new_producer(term, tx_id, pid);
+    });
+}
+
+ss::future<tm_stm::op_status> tm_stm::register_new_producer(
+  model::term_id term,
+  kafka::transactional_id tx_id,
+  model::producer_identity pid) {
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx != _tx_table.end()) {
+        co_return tm_stm::op_status::conflict;
+    }
+
+    auto tx = tm_transaction{
+      .id = tx_id,
+      .pid = pid,
+      .tx_seq = model::tx_seq(0),
+      .status = tm_transaction::tx_status::ongoing,
+    };
+    tx_updated_cmd cmd{.tx = tx, .prev_etag = tx.etag};
+    auto batch = serialize_cmd(cmd, tm_update_batch_type);
+
+    auto r = co_await replicate_quorum_ack(term, std::move(batch));
+
+    if (!r) {
+        co_return tm_stm::op_status::unknown;
+    }
+
+    auto deadline = model::timeout_clock::now()
+                    + config::shard_local_cfg().tm_sync_timeout_ms.value();
+    co_await wait(model::offset(r.value().last_offset()), deadline);
+
+    ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        co_return tm_stm::op_status::conflict;
+    }
+    if (ptx->second.etag != tx.etag) {
+        co_return tm_stm::op_status::conflict;
```

Comment:
but they already return conflict? did you want to comment something else?

Reply:
no my comment stands: I'm asking if return conflict in both cases is like a typo. it didn't fit what looked like a pattern that I saw previously in this file where `not_found` was returned when `tx_id` wasn't found in the transaction table, but here conflict is returned so it was worth asking.

```
    if (ptx == _tx_table.end()) {
         return checked<tm_transaction, tm_stm::op_status>(
           tm_stm::op_status::not_found);
```

Reply:
My reasoning was this: 1) we know that the command was replicated 2) we're at some location in the future after it was applied so if the tx isn't in _tx_table it means that somebody removed it and it's a conflict. But nobody does it at least now so it's an impossible situation. But anyway I've changed it to op_status::unknown

Reply:
Sorry, just to be clear: I don't really understand what the return values mean, with my comment I only wanted to make sure the two return values were intentional because it seemed to not match a previous pattern and may have been a mistake.